blog/2022-05-13-caching-api-requests-with-long-lived-workflows.md

```diff
@@ -0,0 +1,231 @@
+posted_on_: 2022-05-13T00:00:00Z
+slug: caching-api-requests-with-long-lived-workflows
+title: 'Caching API Requests With Long-Lived Workflows in Temporal'
+author: Valeri Karpov
+author_title: Community Member
+author_image_url: https://avatars.githubusercontent.com/u/1620265?v=4
+release_version: V1.15
+---
+
+<!--truncate-->
+
+There is no time limit on [Temporal Workflow Executions](https://docs.temporal.io/docs/concepts/what-is-a-workflow-execution).
+You can write a Workflow that runs forever, storing some state and responding to Signals and Queries that come in, as long as you remember to [Continue As New](https://docs.temporal.io/docs/concepts/what-is-continue-as-new).
+One neat use case for long-lived Workflows is caching API requests.
+
+For example, suppose you want to display prices in different currencies based on cached exchange rates.
+Exchange rate APIs are often expensive for high volumes of requests, so caching can be worthwhile as long as stale data isn't a problem for your use case.
+You can create a single Temporal Workflow that makes 1 API request per day to get the latest exchange rates for a set of currencies, and stores the most recent result, with no explicit database calls or cron jobs.
+In this blog post, I'll describe how you can write an API caching Workflow for the [moneyconvert.net API](https://moneyconvert.net/) with the Temporal TypeScript SDK.
+You can find the [full source code for this example on GitHub](https://github.com/vkarpov15/temporal-api-caching-example).
+
+## Getting Started
+
+When you make an HTTP GET request to https://cdn.moneyconvert.net/api/latest.json, the API returns JSON that looks like what you see below.
+
+```
+{
+  "table": "latest",
+  "rates": {
+    "AED": 3.6731,
+    "AFN": 87.249991,
+    "ALL": 114.227196,
+    "AMD": 473.915796,
+    ...
+  },
+  "lastupdate": "2022-05-11T00:46:04.483000+00:00"
+}
+```
+
+Because Temporal Workflows must be deterministic, you can't make an API request directly from a Workflow.
+You need to create an Activity that makes an API request, and call that Activity from your Workflow.
+Below is an [`activities.ts` file](https://github.com/vkarpov15/temporal-api-caching-example/blob/main/src/activities.ts) that makes an HTTP request to the above API endpoint, and returns the exchange rates.
+
+```ts
+import axios from 'axios';
+
+export async function getExchangeRates(): Promise<any> {
+  const res = await axios.get('https://cdn.moneyconvert.net/api/latest.json');
+  return res.data.rates;
+};
+```
+
+Next, let's write a Workflow that calls this Activity once per day.
+With Temporal Workflows, you can simply write a `while (true) {}` loop with a [JavaScript sleep](https://masteringjs.io/tutorials/fundamentals/sleep) that pauses the Workflow until the next time you need to refresh exchange rates.
+Writing the Workflow this way may seem counterintuitive, because we've all had to learn over the course of our careers that writing applications isn't this easy.
```

Comment:
```suggestion
Writing the Workflow this way might seem counterintuitive, because we've all had to learn over the course of our careers that writing applications isn't this easy.
```